Blacklist for hashing

monchier opened the pull request:
**Issue:** https://github.com/streamlit/streamlit/issues/242

**Description:** Add a blacklist reusing code from the File Watcher. I extracted the code for the blacklist from the FileWatcher. This allows us to have a single point of configuration and some reuse. I added the logic to a class called `BlackList`. Also added tests, mostly moving them out of the FileWatcher tests.

---

**Contribution License Agreement**

By submiting this pull request you agree that all contributions to this project are made under the Apache 2.0 license.

monchier commented on lib/streamlit/hashing.py:
```diff
@@ -290,8 +294,11 @@ def _to_bytes(self, obj, context):
                     return self.to_bytes("%s.%s" % (obj.__module__, obj.__name__))
 
                 h = hashlib.new(self.name)
-                # TODO: This may be too restrictive for libraries in development.
```
If this is an actual limitations, we should link an issue to it. Since I am changing this code, I am removing it.

tvst replied:
Agreed. 

I don't think this is a use-case that matters for Streamlit authors anyway.

monchier commented on lib/tests/streamlit/watcher/LocalSourcesWatcher_test.py:
```diff
@@ -215,35 +214,6 @@ def test_config_blacklist(self, fob):
         config.set_option("server.folderWatchBlacklist", prev_blacklist)
 
-    @patch("streamlit.watcher.LocalSourcesWatcher.FileWatcher")
-    def test_auto_blacklist(self, _):
```
I moved this to `black_list_test`. Let me know if I missed some coverage.

tvst commented on lib/streamlit/black_list.py:
```diff
@@ -0,0 +1,52 @@
+    """
+    Implement a black list object with globbing used to test whether a file is
+    in a folder in the blacklist.
+    """
```
Use Numpy doc style

monchier replied:
Will do

tvst commented on lib/streamlit/black_list.py:
```diff
@@ -0,0 +1,52 @@
+]
+
+
+class BlackList(object):
```
Rename to FolderBlackList

monchier replied:
Sure.

tvst commented on lib/streamlit/black_list.py:
```diff
@@ -0,0 +1,52 @@
+    def _file_is_in_folder(filepath, folderpath_glob):
+        # Strip trailing slash if it exists
+        if folderpath_glob.endswith("/"):
+            folderpath_glob = folderpath_glob[:-1]
```
I have a PR that fixes this function. Please merge in the changes. See #265

monchier replied:
Ok, I can just merge it in. Seems your PR is going to be merged earlier.

tvst commented on lib/streamlit/hashing.py:
```diff
@@ -290,8 +294,10 @@ def _to_bytes(self, obj, context):
-                if os.path.abspath(obj.__code__.co_filename).startswith(os.getcwd()):
+                filepath = os.path.abspath(obj.__code__.co_filename)
+                if filepath.startswith(
+                    os.getcwd()
```
I have to think more deeply about this, but just something I noticed: this code uses `cwd` whereas `LocalSourcesWatcher` uses the folder where the main script lives.

Does it make sense to unify the two behaviors? If so, which is best: cwd or main folder?

monchier replied:
I wonder this is just an artifact of the code being written by two different people. I do think there might be issues with both. What if the main script is in a subdirectory? And what if we execute streamlit from a path that does not include CWD? Should be document and make this a user-option. We can make clear that only a subset of files will be watched/checked for mutation.

I like this to be the same for FileWatcher and here.

tvst commented on lib/streamlit/black_list.py:
```diff
@@ -0,0 +1,52 @@
+        )
+
+    @staticmethod
+    def _file_is_in_folder(filepath, folderpath_glob):
```
Move this to a utility file, so it can be reused in LocalSourcesWatcher

monchier replied:
`BlackList` would be shared, so we should not need to move this to a utility.

monchier commented:
@tvst Let's chat whether we want this to be shared with FileWatcher. Let's also clarify the role of CWD/main folder path

monchier commented:
> @tvst Let's chat whether we want this to be shared with FileWatcher. Let's also clarify the role of CWD/main folder path

We decided to share this with FileWatcher and to use the main script folder.

monchier commented on lib/streamlit/hashing.py:
```diff
@@ -296,8 +315,11 @@ def _to_bytes(self, obj, context):
+
+                if util.file_is_in_folder(
+                    filepath, _get_main_script_directory()
+                ) and not self._folder_black_list.is_blacklisted(filepath):
```
I am separately checking 
- whether the file is in the current directory
- whether the file is not in the blacklist

monchier commented on lib/streamlit/watcher/LocalSourcesWatcher.py:
```diff
@@ -158,16 +159,13 @@ def update_watched_modules(self):
-                file_is_local = _file_is_in_folder(filepath, self._report.script_folder)
+                file_is_local = util.file_is_in_folder(
+                    filepath, self._report.script_folder
+                )
```
Same as `hashing.py`, I am leaving i) checking if the file is local and ii) checking against the blacklist separate.
Also, here we use then ` self._report.script_folder` as the main script folder.

monchier commented on lib/streamlit/hashing.py:
```diff
@@ -156,7 +159,19 @@ def _hashing_error_message(start):
+    import os
+
+    main_path = __main__.__file__
+    return os.path.dirname(main_path)
```
I am assuming this happens while I am executing a streamlit script and this is well defined.

monchier commented on lib/tests/streamlit/util_test.py:
```diff
@@ -179,3 +179,8 @@ def test_get_external_ip(self):
+    def test_file_is_in_folder(self):
+        self.assertTrue(util.file_is_in_folder("/user/name/test", "/user"))
+        self.assertTrue(util.file_is_in_folder("/user/name/test", "/user/*"))
+        self.assertFalse(util.file_is_in_folder("/user/name/test", "/user/other"))
```
basic tests for this function.

monchier commented:
I did not include the check for a file to be local in the blacklist. It seems to be it does not belong there semantically and would be a bit opaque, but there is some replication since the two checks are in the File Watcher and in `hashing.py` is a slightly different format. Also I moved `file_is_in_folder` to `util.py` and moved tests - it could have be together with the blacklist and be exposed as a public method, but does not belong to the same file and it would be odd to say "black_list.file_is_in_folder", so `util` seemed a better place.

tvst commented on lib/streamlit/folder_black_list.py:
```diff
@@ -30,9 +29,8 @@ class FolderBlackList(object):
-    The blacklist does blacklist anything that is not in the directory of the
-    main script and anything that matches the `DEFAULT_FOLDER_BLACKLIST`.
+    The blacklist does blacklist anything that matches the
+    `DEFAULT_FOLDER_BLACKLIST`.
```
Nit: change comment to 
```
Blacklist any path that matches a glob in `DEFAULT_FOLDER_BLACKLIST`.
```

tvst commented on lib/streamlit/hashing.py:
```diff
@@ -408,3 +404,13 @@ def _code_to_bytes(self, code, context):
+        import __main__
+        import os
+
+        main_path = __main__.__file__
```
Does this point to the right file? In some toy tests on my computer, it seems to point to the equivalent of `cli.py`

Here's what I did:

```python
# main_test.py

with open('main_subtest.py') as script_file:
    script = script_file.read()

exec(script, {})
```

```python
# main_subtest.py

import streamlit as st
import __main__
st.write(__main__.__file__)
```

Expected: should see "main_subtest.py" in Streamlit app
Actual: saw "main_test.py" in Streamlit app

monchier replied:
I tested this in `streamlit`... In the case below, it shows the correct path. Let me try to repro. I would rather use `_report.script_path` but we need to pass it down to the hasher, which I could do, but I need to change the interface.

<img width="756" alt="Screen Shot 2019-10-08 at 11 42 32 AM" src="https://user-images.githubusercontent.com/6821233/66423647-f1216100-e9c0-11e9-8b9a-a48abe404bb7.png">

monchier replied:
Interesting:
https://github.com/streamlit/streamlit/blob/develop/lib/streamlit/ScriptRunner.py#L293

We do this, before running: 
```
sys.modules["__main__"] = module
```

monchier replied:
Actually, happens here:```
module.__dict__["__file__"] = self._report.script_path
```

monchier replied:
Which is what we are looking for, right?
I think if we like this solution we need at the very least a test.

tvst replied:
You're right. This code is fine, then!

Can you add a comment before line 415 explaining why this is fine?

monchier replied:
Will add the comment.

tvst commented on lib/streamlit/util.py:
```diff
@@ -451,3 +452,26 @@ def is_namedtuple(x):
     return platform.system() == "Darwin"
+
+
+def file_is_in_folder(filepath, folderpath_glob):
```
Rename to `file_is_in_folder_glob` or something like that, for clarity.
